### DISCUSSION TITLE
Java instrumentation and clock skew problem

### DISCUSSION THREAD

**Discussion by rkondrashov** [open-telemetry/opentelemetry-java-instrumentation#2590] [Q&A]
In k8s environment I frequently see some wrong aligned spans generated by java instrumentation, but the spans generated by other instrumentations like go and python are always aligned correctly.

![jaeger-otel-k8s-clock-skew](https://user-images.githubusercontent.com/12693997/111466284-64391600-8723-11eb-8967-a2eea26beecb.png)

I've created a [sample project](https://github.com/rkondrashov/opentelemetry-playground-go-java) to reproduce this behavior and tested different combinations of the go and java applications on my local machine. The behavior is the same: the spans generated by java instrumentation are often wrong aligned and the spans generated by go instrumentation are not affected by this problem.

![jaeger-otel-go-java-clock-skew-1](https://user-images.githubusercontent.com/12693997/111466386-83d03e80-8723-11eb-87b0-93dc3e67ac0b.png)

![jaeger-otel-go-java-clock-skew-2](https://user-images.githubusercontent.com/12693997/111466389-8468d500-8723-11eb-90a5-2891a13ad208.png)

![jaeger-otel-go-java-clock-skew-3](https://user-images.githubusercontent.com/12693997/111466393-859a0200-8723-11eb-8753-fab0caf857ff.png)

Is it a kind of the clock skew problem due to the different programming languages and time libraries or an issue with java instrumentation? For me it looks like a time accuracy problem in java instrumentation, an application which calls itself (same host, jre, agent and libraries) also produces wrong aligned spans.

![jaeger-otel-java-clock-skew](https://user-images.githubusercontent.com/12693997/111466409-8cc11000-8723-11eb-99fd-34518ebe3300.png)

[traces.zip](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/6156412/traces.zip)

**Comment by trask** [comment]
hey @rkondrashov, thanks for the report!

For the interop tests with other languages, I believe this is the issue you are seeing: #2600

On the last image (where only java is present), is the red box below your concern?

![image](https://user-images.githubusercontent.com/218610/111728167-90d75580-8829-11eb-9268-dfd0ddca6e0f.png)

If so, there are a couple possible explanations for that
* one is that it looks like your test application is using Spring RestTemplate which defaults to using HttpURLConnection, and that instrumentation only captures time-to-first-byte, which is why that line could be a bit short. You could try configuring Spring RestTemplate to use OkHttp or Apache HttpClient, as the javaagent instrumentation should capture time-to-last-byte for both of those.
* another is that the server span captures the full servlet/container request-response lifecycle, which is not exactly tied to returning of the response, so can go slightly beyond the returning of the response.

**Comment by rkondrashov** [comment]
@trask thanks for the detailed answer. 

That's right, on the last screenshot (java app only) my concern is about the end-time of the request in the red box below. It looks a bit curious, so that on a synchronous request the client receives a response before the server returns it. The question, is it reasonable to measure the entire request-response lifecycle duration of the servlet/container or just the response time instead. I mean, the end users are usually only interested in the request execution time.
 
I'll test it with other suggested clients and post the results.

**Comment by rkondrashov** [comment]
I have tested it with OkHttp and Apache clients and other containers like Jetty and Netty (WebFlux). The result is the same, the spans from time to time are also wrong aligned.

![jaeger-otel-java-okhttp-clock-skew](https://user-images.githubusercontent.com/12693997/112630596-15cbfb80-8e36-11eb-84de-caa4d209d29f.png)

![jaeger-otel-java-apache-clock-skew](https://user-images.githubusercontent.com/12693997/112630613-20869080-8e36-11eb-9e11-1bd9e6fd6a12.png)

![jaeger-otel-java-jetty-clock-skew](https://user-images.githubusercontent.com/12693997/112630622-25e3db00-8e36-11eb-9cb0-5e785be2b0a4.png)

![jaeger-otel-java-webflux-clock-skew](https://user-images.githubusercontent.com/12693997/112630635-2b412580-8e36-11eb-8cf5-a82b89a0dd28.png)

What I noticed was that the deviation is always in the microseconds range, mostly between 100 and 700 microseconds. Could it be also caused by [#2600](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2600)?
